### PR TITLE
Bosh backup

### PR DESCRIPTION
**Pull request by szaouam**
This an internal orange review, once merged, on orange-cloudfoundry branch, we'll send a new PR to the upstream repo . Following is the planned PR text to be submitted to the community:

---- 

changes:
- Enables usage of S3-compatible API endpoints others than ``s3.amazonaws.com``
- add openstack manifest for bosh v1 and v2 syntaxes (for collocated and remote  deployment)
- add proxy support for backup and restore processes
- add lock to avoid simultaneous dumps of a same db
- enable the debug mode for backup and restore 
- update specs (remove default value for some properties)
   ==> pgbackup.username, pgbackup.password
           pgbackup.s3.access_key_id, pgbackup.s3.secret_access_key
- modify  docs/pgbackup.md to add the "Openstack Deployment"

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,77 +1,46 @@
...
-```sh
-mkdir -p ~/workspace/manifests
-bosh download manifest pgbackup-development ~/workspace/manifests/pgbackup.yml
-bosh deployment ~/workspace/manifests/pgbackup.yml
-```
+Next edit the manifest file provided with the release (manifest/postgres-backup.yml).
 
-Alternatively, you can make your manifest. For example to prepare a manifest for
-bosh-lite (warden) using the 'centos' stemcell we would do the following:
+This release aims to support bosh v2 manifest. Currently, only Openstack templates are offered by default,
```
Comment (gberche-orange) [inline README.md]: This release aims to support both v1 and v2 manifests

File: jobs/pgbackup/spec [hunk 2]
```diff
@@ -33,24 +29,60 @@ properties:
...
+    default: "PLEASE-CHANGE-THIS"
   pgbackup.pgdump.arguments:
     description: "Extra arguments to pass to PostgreSQL's pg_dump utility"
     default: ""
 
+
+  pgbackup.proxy.http.active:
+    description: "set to true if you want to use a proxy.  false is the default value"
+    default: "false"
+  pgbackup.proxy.http.host:
```
Comment (gberche-orange) [inline jobs/pgbackup/spec]: naming, consider `` pgbackup.s3.http_proxy_host``  ?

File: jobs/pgbackup/spec [hunk 3]
```diff
@@ -33,24 +29,60 @@ properties:
...
+
+
   pgbackup.schedule.minute:
     description: "Backup schedule - minute (00-59)"
-    default: "00"
+    default: "01"
   pgbackup.schedule.hour:
     description: "Backup schedule - hour (00-23)"
-    default: "01"
+    default: "*"
```
Comment (gberche-orange) [inline jobs/pgbackup/spec]: Consider restoring daily default

File: jobs/pgbackup/spec [hunk 4]
```diff
@@ -74,5 +106,5 @@ properties:
     description: "Secret Access key for storing backups in Amazon S3"
     default: ""
```
Comment (gberche-orange) [inline jobs/pgbackup/spec:75]: consider removing the default to make the property mandatory

File: jobs/pgbackup/templates/bin/pgbackup [hunk 5]
```diff
@@ -27,48 +22,73 @@ when 9.1, "9.1"
 when 9.0, "9.0"
   "postgres90"
 else
-  "postgres95"
+  "postgres90"
```
Comment (gberche-orange) [inline jobs/pgbackup/templates/bin/pgbackup]: consider restoring 9.5 default (this release is not specific to bosh director). Likely the pgdump/pgrestore are compatible with an old pg db
* CF is 9.4
* Bosh is 9.0

File: jobs/pgbackup/templates/bin/pgbackup [hunk 6]
```diff
@@ -27,48 +22,73 @@ when 9.1, "9.1"
...
 end
 %>
 
 PGVERSION="<%= pgversion %>"
-export PGHOST PGPORT PGUSER PGPASSWORD PGVERSION
-echo "export PGHOST=$PGHOST"
-echo "export PGPORT=$PGPORT"
-echo "export PGUSER=$PGUSER"
-echo "export PGVERSION=$PGVERSION"
```
Comment (gberche-orange) [inline jobs/pgbackup/templates/bin/pgbackup:39]: consider restoring logs, and refine with AWS S3 bucket and endpoinbt (but not the secret key)

for the existing debugging property consider invoking a "set -x"

File: jobs/pgbackup/spec [hunk 7]
```diff
@@ -62,17 +59,40 @@ properties:
     default: "*"
 
   pgbackup.s3.bucket:
-    description: "Amazon S3 bucket to store backups in"
-    default: "pgbackup"
+    description: "S3 bucket to store backups in"
   pgbackup.s3.path:
     description: "Path to prepend to backup filenames when storing in Amazon S3"
-    default: ""
+    default: "\"
```
Comment (gberche-orange) [inline jobs/pgbackup/spec]: will this work on S3 or is this default specific to orange S3 blobstore ?

File: jobs/pgbackup/spec [hunk 8]
```diff
@@ -62,17 +59,40 @@ properties:
...
     description: "Secret Access key for storing backups in Amazon S3"
-    default: ""
   pgbackup.s3.endpoint:
     description: "endpoint for backup/restore"
-    default: "s3.amazonaws.com"
+
+
+ pgbackup.s3.proxy.http.host:
+    description: "Hostname or Ip address of the proxy server, http://your_proxy_server"
+    default: ""
```
Comment (gberche-orange) [inline jobs/pgbackup/spec:82]: Do we have use cases for separating http and https proxies ? If not consider a single set of properties for configuring both http and https

File: jobs/pgbackup/spec [hunk 9]
```diff
@@ -62,17 +59,40 @@ properties:
...
+
+  pgbackup.s3.proxy.https.host:
+    description: "Hostname or Ip address of the proxy server, http(s)://your_proxy_server"
+    default: ""
+  pgbackup.s3.proxy.https.port:
+    description: "Port used to achieve the proxy server"
+    default: "3128"
+  pgbackup.s3.proxy.https.user:
+    description: "Username for the proxy server"
```
Comment (gberche-orange) [inline jobs/pgbackup/spec]: Username used in http proxy basic authentication

File: jobs/pgbackup/templates/bin/pgbackup [hunk 10]
```diff
@@ -47,9 +85,27 @@ export AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY
...
-S3_ENDPOINT= "<%= p('pgbackup.s3.endpoint') %>"
+S3_ENDPOINT="<%= p('pgbackup.s3.endpoint') %>"
+
+echo "export PGHOST=$PGHOST"
+echo "export PGPORT=$PGPORT"
+echo "export PGUSER=$PGUSER"
+echo "export PGVERSION=$PGVERSION"
+echo "use S3_ENDPOINT=$S3_ENDPOINT"
+
+# activate debug mode will enable storing password in logs file,
```
Comment (gberche-orange) [inline jobs/pgbackup/templates/bin/pgbackup]: "activate debug mode implies storing passwords in logs file,"

File: jobs/pgbackup/templates/bin/pgbackup [hunk 11]
```diff
@@ -62,13 +118,17 @@ backup_databases() {
...
 
 echo "pgbackup starting up"
-# FIXME: acquire a lock to ensure we aren't running simultaneous dumps
 
-backup_databases
+# if a backup.lock file exist then we assume that an other dump is running, so we will retry to backup in the next run
+if [ ! -f "/tmp/backup.lock" ]; then
+    backup_databases
+else
+    echo "ERROR : an other dump is running or the last backup failed , see logs for more details"
```
Comment (gberche-orange) [inline jobs/pgbackup/templates/bin/pgbackup]: lock is released when backup fails, should read "another dump is still running, skipping this backup, see logs for more details"

File: jobs/pgbackup/templates/bin/pgbackup [hunk 12]
```diff
@@ -62,13 +118,17 @@ backup_databases() {
...
 	fi
+	rm -f "/tmp/backup.lock"
 }
 
 echo "pgbackup starting up"
-# FIXME: acquire a lock to ensure we aren't running simultaneous dumps
 
-backup_databases
+# if a backup.lock file exist then we assume that an other dump is running, so we will retry to backup in the next run
+if [ ! -f "/tmp/backup.lock" ]; then
```
Comment (gberche-orange) [inline jobs/pgbackup/templates/bin/pgbackup:123]: should probably stop waiting on the lock after a max time, such as 1 day, and clear the backup lock ?

**Review by gberche-orange — CHANGES_REQUESTED** [review]
Thanks @szaouam for the improvements, some related comments that might require some changes

File: jobs/pgbackup/spec [hunk 13]
```diff
@@ -60,39 +63,27 @@ properties:
 
   pgbackup.s3.bucket:
     description: "S3 bucket to store backups in"
+    default: "pgbackup"
   pgbackup.s3.path:
     description: "Path to prepend to backup filenames when storing in Amazon S3"
-    default: "\"
```
Comment (szaouam) [inline jobs/pgbackup/spec]: removed : Minor type

File: jobs/pgbackup/spec [hunk 14]
```diff
@@ -60,39 +63,27 @@ properties:
...
-    description: "Username for the proxy server"
-    default: ""
-  pgbackup.s3.proxy.http.password:
-    description: "Password for the proxy server username"
-    default: ""
-
-  pgbackup.s3.proxy.https.host:
-    description: "Hostname or Ip address of the proxy server, http(s)://your_proxy_server"
+  pgbackup.s3.proxy.user:
+    description: "Username used in http proxy basic authentication"
```
Comment (szaouam) [inline jobs/pgbackup/spec]: add a single set of properties for configuring both http and https

File: jobs/pgbackup/templates/bin/pgbackup [hunk 15]
```diff
@@ -5,45 +5,24 @@ set +e # do not want to auto-die - instead, log and continue
...
 #--- export proxy parameters
-if [[ $http_proxy_host != "" ]]; then
-export http_proxy="$http_proxy"
-fi
-if [[ $https_proxy_host != "" ]]; then
-export https_proxy="$https_proxy"
+if [[ $proxy_host != "" ]]; then
+export http_proxy="$proxy"
+export https_proxy="$proxy"
 fi
```
Comment (szaouam) [inline jobs/pgbackup/templates/bin/pgbackup:25]: update proxy configuration to use a single set of parameters

File: jobs/pgbackup/templates/bin/pgbackup [hunk 16]
```diff
@@ -118,16 +108,23 @@ backup_databases() {
...
 	rm -f "/tmp/backup.lock"
 }
 
 echo "pgbackup starting up"
 
+
+# if the backup process fail the .lock file will be not removed.
+# We wait until tomorrow to automatically release the lock
+daily_clear_lock
+
```
Comment (szaouam) [inline jobs/pgbackup/templates/bin/pgbackup:121]: We stop waiting on the lock after 1 day

**Review by szaouam — COMMENTED** [review]
Update release (see comments)

File: config/blobs.yml [hunk 17]
```diff
@@ -35,3 +35,7 @@ postgres95/postgresql-9.5beta1.tar.gz:
   object_id: 2c204877-0297-48e6-901e-bf5331442a94
   sha: 6a5fe49efb50e235044dc9abdf4c12c423696d0d
   size: 23836877
+gof3r/gof3r-0.5.0.tar.gz:
+  object_id: 1591f857-2df5-4ac0-911d-88e86177543f
+  sha: 9cbf1191604d66a174b078e54c20ea9ab1685742
+  size: 2555335
```
Comment (szaouam) [inline config/blobs.yml]: update the gof3r package

File: packages/gof3r/spec [hunk 18]
```diff
@@ -2,4 +2,4 @@
 name: gof3r
 dependencies: []
 files:
-  - gof3r/gof3r-*
+  - gof3r/gof3r-0.5.0.tar.gz
```
Comment (szaouam) [inline packages/gof3r/spec]: update the gof3r package

**Review by szaouam — COMMENTED** [review]
downgrade the gof3r package version to 0.4.10  due to compatibility with current release

File: pgbackup-dev [hunk 19]
```diff
@@ -31,7 +31,7 @@ usage() {
   echo "
 Usage: $0 <prepare|blobs|manifest|release|destroy> [options]
 Where [options] for the 'prepare', 'manifest' and 'stemcell' actions are:
-  'warden' or 'aws-ec2'
+  'warden' or 'aws-ec2' or 'openstack'
 "
```
Comment (szaouam) [inline pgbackup-dev:35]: update help message

File: pgbackup-dev [hunk 20]
```diff
@@ -115,6 +115,20 @@ prepare_stemcell() {
...
+      hypervisor="kvm"
+      case ${stemcellOS} in
+        (ubuntu)
+          stemcellURL="https://bosh.io/d/stemcells/bosh-openstack-kvm-ubuntu-trusty-go_agent"
+          ;;
+        (*)
+          fail "Unknown Stemcell OS '${stemcellOS}'"
+          ;;
+      esac
+      ;;
```
Comment (szaouam) [inline pgbackup-dev:131]: add openstack stemcells

**Review by szaouam — COMMENTED** [review]
add openstack support for bosh lite

File: jobs/pgbackup/spec [hunk 21]
```diff
@@ -62,17 +62,28 @@ properties:
...
     default: ""
   pgbackup.s3.access_key_id:
     description: "Access Key ID for storing backups in Amazon S3"
-    default: ""
   pgbackup.s3.secret_access_key:
     description: "Secret Access key for storing backups in Amazon S3"
-    default: ""
   pgbackup.s3.endpoint:
-    description: "endpoint for backup/restore"
-    default: "s3.amazonaws.com"
```
Comment (gberche-orange) [inline jobs/pgbackup/spec:78]: Please restore default value as to avoid breaking compatibility of existing deployments

File: jobs/pgbackup/spec [hunk 22]
```diff
@@ -62,17 +62,28 @@ properties:
...
+
+
+  pgbackup.s3.proxy.host:
+    description: "Hostname or Ip address of the proxy server, http://your_proxy_server"
+    default: ""
+  pgbackup.s3.proxy.port:
+    description: "Port used to achieve the proxy server"
+    default: "3128"
+  pgbackup.s3.proxy.user:
+    description: "Username used in http proxy basic authentication"
```
Comment (gberche-orange) [inline jobs/pgbackup/spec]: ``Username used in http proxy basic authentication. Setting an empty string will skip proxy authentication (this is the default)."

File: jobs/pgbackup/spec [hunk 23]
```diff
@@ -62,17 +62,28 @@ properties:
...
+    default: ""
+  pgbackup.s3.proxy.password:
+    description: "Password used in http proxy basic authentication"
```
Comment (gberche-orange) [inline jobs/pgbackup/spec]: "Password used in http proxy basic authentication. Setting an empty string will skip proxy authentication (this is the default)."

File: jobs/pgbackup/spec [hunk 24]
```diff
@@ -76,3 +76,16 @@ properties:
   pgbackup.s3.endpoint:
     description: "endpoint for backup/restore"
     default: "s3.amazonaws.com"
+
+  pgbackup.s3.proxy.host:
+    description: "Hostname or Ip address of the proxy server, http://your_proxy_server"
```
Comment (gberche-orange) [inline jobs/pgbackup/spec]: "Hostname or Ip address of the proxy server, http://your_proxy_server. Set an empty string to not configure any proxy (this is the default).

**Review by gberche-orange — CHANGES_REQUESTED** [review]
Please consider comments inline for the spec file wording

**Review by gberche-orange — APPROVED** [review]

**Pull request merged**